subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt

```diff
@@ -352,6 +353,8 @@ class CompileKotlinScript(
     override fun visitIdentityInputs(
         visitor: InputVisitor
     ) {
+        val assignmentOverloadEnabled = KotlinDslAssignment.isAssignmentOverloadEnabled()
```

Comment:
:x: Let's get this from `ProgramId`, and avoid going directly to `KotlinDslAssignment`.

🤔 We can also get some of the other parameters from `ProgramId` while at it.

Reply:
Changed to ProgramId. Although this change is a bit more complex so I did it in a separate commit:
https://github.com/gradle/gradle/pull/22240/commits/16de9c69adb8f5e4592960fa7e4f1de937826b0e.

Some tests are a bit harder to write due to this change, so if we remove the assignment opt-in flag maybe we should revert this change.